src/charm.py

```diff
@@ -125,9 +156,146 @@ def _configure_nrf(self, event: EventBase) -> None:
 
+    def _on_certificates_relation_created(self, event: EventBase) -> None:
+        """Generates Private key."""
+        if not self.unit.is_leader():
+            return
+        if not self._container.can_connect():
+            event.defer()
+            return
+        self._generate_private_key()
+
+    def _on_certificates_relation_broken(self, event: EventBase) -> None:
+        """Deletes TLS related artifacts and reconfigures workload."""
+        if not self.unit.is_leader():
+            return
+        if not self._container.can_connect():
+            event.defer()
+            return
+        self._delete_private_key()
+        self._delete_csr()
+        self._delete_certificate()
+        self._configure_nrf(event)
+
+    def _on_certificates_relation_joined(self, event: EventBase) -> None:
+        """Generates CSR and requests new certificate."""
+        if not self.unit.is_leader():
+            return
+        if not self._container.can_connect():
+            event.defer()
+            return
+        if not self._private_key_is_stored():
+            event.defer()
+            return
+        self._request_new_certificate()
+
+    def _on_certificate_available(self, event: CertificateAvailableEvent) -> None:
+        """Pushes certificate to workload and configures workload."""
+        if not self.unit.is_leader():
+            return
+        if not self._container.can_connect():
+            event.defer()
+            return
+        if not self._csr_is_stored():
+            logger.warning("Certificate is available but no CSR is stored")
+            return
+        if event.certificate_signing_request != self._get_stored_csr():
+            logger.debug("Stored CSR doesn't match one in certificate available event")
+            return
+        self._store_certificate(event.certificate)
+        self._configure_nrf(event)
+
+    def _on_certificate_expiring(self, event: CertificateExpiringEvent) -> None:
+        """Requests new certificate."""
+        if not self.unit.is_leader():
+            return
+        if not self._container.can_connect():
+            event.defer()
+            return
+        if event.certificate != self._get_stored_certificate():
+            logger.debug("Expiring certificate is not the one stored")
+            return
+        self._request_new_certificate()
+
+    def _generate_private_key(self) -> None:
+        """Generates and stores private key."""
+        private_key = generate_private_key()
+        self._store_private_key(private_key.decode())
+
+    def _request_new_certificate(self) -> None:
+        """Generates and stores CSR, and uses to request new certificate."""
+        private_key = self._get_stored_private_key()
+        csr = generate_csr(
+            private_key=private_key.encode(),
+            subject=CERTIFICATE_COMMON_NAME,
+            sans_dns=[CERTIFICATE_COMMON_NAME],
+        )
+        self._store_csr(csr.decode().strip())
```

Comment:
Same thing with `decode()` and `strip()`.

Reply:
Done